src/jit/hwintrinsicxarch.cpp

```diff
@@ -234,6 +234,57 @@ int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic)
     return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
 }
 
+//------------------------------------------------------------------------
+// numArgsOfHWIntrinsic: get the number of arguments based on table and
+// if numArgs is -1 check number of arguments using GenTreeHWIntrinsic node
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function
+//    node      -- GenTreeHWIntrinsic* node
+//
+// Return Value:
+//     number of arguments
+//
+int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic, GenTreeHWIntrinsic* node)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+
+    int numArgs = hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
+    if (numArgs != -1)
+    {
+        return numArgs;
+    }
+
+    assert(numArgs == -1);
+
+    GenTree* op1 = node->gtGetOp1();
+    GenTree* op2 = node->gtGetOp2();
+
+    if (op2 != nullptr)
+    {
+        return 2;
+    }
+
+    if (op1 != nullptr)
+    {
+        if (op1->OperIsList())
+        {
+            assert(op1->AsArgList()->Rest()->Rest()->Current() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest()->Rest() == nullptr);
+            return 3;
```

Comment:
I would still like to actually enumerate the `argList`, as it is both future proof (and will be required to support future ISAs) and is more correct in the case of the assertions being wrong.

Reply:
It also won't matter, perf wise, since the majority of cases will be table driven.